src/storage/aggregator.cc

```diff
@@ -0,0 +1,383 @@
+        }
+        bool ok = UpdateAggrVal(base_row_view_, row_ptr, &aggr_buffer);
+        if (!ok) {
+            PDLOG(ERROR, "Update aggr value failed");
+            return false;
+        }
+    }
+    return true;
+}
+
+bool Aggregator::GetAggrBuffer(const std::string& key, AggrBuffer* buffer) {
+    std::lock_guard<std::mutex> lock(mu_);
+    if (aggr_buffer_map_.find(key) == aggr_buffer_map_.end()) {
+        return false;
+    }
+    *buffer = aggr_buffer_map_.at(key).buffer_;
+    return true;
+}
+
+bool Aggregator::GetAggrBufferFromRowView(const codec::RowView& row_view, const int8_t* row_ptr, AggrBuffer* buffer) {
+    if (buffer == nullptr) {
+        return false;
+    }
+    row_view.GetValue(row_ptr, 1, DataType::kTimestamp, &buffer->ts_begin_);
+    row_view.GetValue(row_ptr, 2, DataType::kTimestamp, &buffer->ts_end_);
+    row_view.GetValue(row_ptr, 3, DataType::kInt, &buffer->aggr_cnt_);
+    char* ch = NULL;
+    uint32_t ch_length = 0;
+    row_view.GetValue(row_ptr, 4, &ch, &ch_length);
+    switch (aggr_col_type_) {
+        case DataType::kSmallInt:
+        case DataType::kInt:
+        case DataType::kBigInt: {
+            int64_t origin_val = *reinterpret_cast<int64_t*>(ch);
+            buffer->aggr_val_.vlong = origin_val;
+            break;
+        }
+        case DataType::kFloat: {
+            float origin_val = *reinterpret_cast<float*>(ch);
+            buffer->aggr_val_.vfloat = origin_val;
+            break;
+        }
+        case DataType::kDouble: {
+            double origin_val = *reinterpret_cast<double*>(ch);
+            buffer->aggr_val_.vdouble = origin_val;
+            break;
+        }
+        default: {
+            PDLOG(ERROR, "Unsupported data type");
+            return false;
+        }
+    }
+    return true;
+}
+
+bool Aggregator::FlushAggrBuffer(const std::string& key, const AggrBuffer& buffer) {
+    std::string encoded_row;
+    std::string aggr_val;
+    switch (aggr_col_type_) {
+        case DataType::kSmallInt:
+        case DataType::kInt:
+        case DataType::kBigInt: {
+            int64_t tmp_val = buffer.aggr_val_.vlong;
+            aggr_val.assign(reinterpret_cast<char*>(&tmp_val), sizeof(int64_t));
+            break;
+        }
+        case DataType::kFloat: {
+            float tmp_val = buffer.aggr_val_.vfloat;
+            aggr_val.assign(reinterpret_cast<char*>(&tmp_val), sizeof(float));
+            break;
+        }
+        case DataType::kDouble: {
+            double tmp_val = buffer.aggr_val_.vdouble;
+            aggr_val.assign(reinterpret_cast<char*>(&tmp_val), sizeof(double));
+            break;
+        }
+        default: {
+            PDLOG(ERROR, "Unsupported data type");
+            return false;
+        }
+    }
+
+    int str_length = key.size() + aggr_val.size();
+    uint32_t row_size = row_builder_.CalTotalLength(str_length);
+    encoded_row.resize(row_size);
+    {
+        std::lock_guard<std::mutex> lock(rb_mu_);
```

Comment:
Remove the lock. reuse `RowBuilder` if #1525 has been done

Reply:
yep, it will be done in the next pr.